### ISSUE TITLE
The behavior of `--no-explicit-reexport` is inconsistent/misleading

### ISSUE THREAD

**Issue by pavel-kirienko** [python/mypy#11706]
**Bug Report**

The behavior of `--no-explicit-reexport` is inconsistent/misleading, see MWE below.

**To Reproduce**

Define the following package of four files:

```
├── a
│   ├── __init__.py
│   └── X.py
├── c.py
└── d.py
```

**`a/__init__.py`**:

```python
from a.X import X as X
Y = X
```

**`a/X.py`**

```python
class X:
    pass
```

**`c.py`**

```python
from a import Y as W
```

**`d.py`**

```python
from c import W
```

**Expected Behavior**

```bash
$ mypy --strict --no-implicit-reexport .
Success: no issues found in 4 source files
```

**Actual Behavior**

```bash
$ mypy --strict --no-implicit-reexport .
d.py:1: error: Module "c" does not explicitly export attribute "W"; implicit reexport disabled
Found 1 error in 1 file (checked 4 source files)
```

The docs state that the form `import ... as ...` is sufficient to mark the imported item for re-export, so this is expected to work. However, the most suspicious part is that if you modify `c.py` (sic!) as follows:

```diff
-- from a import Y as W
++ from a import X as W
```

then MyPy reports no error.

One would expect both cases to be accepted.

Note that if `a` is not a package, both cases yield a re-export error.

**Your Environment**

- Mypy version used: 0.910
- Mypy command-line flags: `--strict --no-implicit-reexport`
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.10
- Operating system and version: GNU/Linux Manjaro 21.2

**Comment by hauntsaninja** [comment]
Thanks, can repro the inconsistency on master, the expected behaviour is an error in both cases. This was somewhat recently clarified in PEP 484 (https://github.com/python/peps/pull/1630/files)

**Comment by hauntsaninja** [comment]
I'd welcome a corresponding update to our docs, if you want to make a PR for it! :-)

**Comment by pavel-kirienko** [comment]
>Thanks, can repro the inconsistency on master, the expected behaviour is an error in both cases. This was somewhat recently clarified in PEP 484

This is a bit unexpected. What is the recommended pattern for re-exporting an entity under a different name? The following obvious approach looks verbose:

```python
from x import y
z = y
```

**Comment by hauntsaninja** [comment]
Yup, that's what we used in typeshed :-)

**Comment by hauntsaninja** [comment]
Okay, I made #11707 to fix the inconsistency and it looks like #11083 updated the docs already
